File: native/cocos/editor-support/MeshBuffer.cpp
/****************************************************************************
 Copyright (c) 2020-2023 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights to
 use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 of the Software, and to permit persons to whom the Software is furnished to do so,
 subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
****************************************************************************/

#include "MeshBuffer.h"
#include "2d/renderer/UIMeshBuffer.h"

MIDDLEWARE_BEGIN

static const ccstd::vector<gfx::Attribute> ATTRIBUTES_V3F_T2F_C4B{
    gfx::Attribute{gfx::ATTR_NAME_POSITION, gfx::Format::RGB32F},
    gfx::Attribute{gfx::ATTR_NAME_TEX_COORD, gfx::Format::RG32F},
    gfx::Attribute{gfx::ATTR_NAME_COLOR, gfx::Format::RGBA8, true},
};

static const ccstd::vector<gfx::Attribute> ATTRIBUTES_V3F_T2F_C4B_C4B{
    gfx::Attribute{gfx::ATTR_NAME_POSITION, gfx::Format::RGB32F},
    gfx::Attribute{gfx::ATTR_NAME_TEX_COORD, gfx::Format::RG32F},
    gfx::Attribute{gfx::ATTR_NAME_COLOR, gfx::Format::RGBA8, true},
    gfx::Attribute{gfx::ATTR_NAME_COLOR2, gfx::Format::RGBA8, true},
};

MeshBuffer::MeshBuffer(int vertexFormat)
: MeshBuffer(vertexFormat, INIT_INDEX_BUFFER_SIZE, MAX_VERTEX_BUFFER_SIZE) {
}

MeshBuffer::MeshBuffer(int vertexFormat, size_t indexSize, size_t vertexSize)
: _vertexFormat(vertexFormat), _ib(indexSize), _vb(vertexSize * vertexFormat * sizeof(float)) {
    _vb.setMaxSize(MAX_VERTEX_BUFFER_SIZE * _vertexFormat * sizeof(float));
    _ib.setMaxSize(100 * INIT_INDEX_BUFFER_SIZE);
    _vb.setFullCallback([this] {
        uploadVB();
        uploadIB();
        _vb.reset();
        _ib.reset();
        next();
    });

    init();
}

MeshBuffer::~MeshBuffer() {
    clear();
}

void MeshBuffer::clear() {
    size_t num = _vbArr.size();
    for (size_t i = 0; i < num; i++) {
        delete _ibArr[i];
        delete _vbArr[i];
    }
    _ibArr.clear();
    _vbArr.clear();
    cleanUIMeshBuffer();
}

void MeshBuffer::init() {
    auto *rIB = new IOTypedArray(se::Object::TypedArrayType::UINT16, _ib.getCapacity());
    _ibArr.push_back(rIB);

    auto *rVB = new IOTypedArray(se::Object::TypedArrayType::FLOAT32, _vb.getCapacity());
    _vbArr.push_back(rVB);
    cleanUIMeshBuffer();
    addUIMeshBuffer();
}

void MeshBuffer::uploadVB() {
    auto length = _vb.length();
    if (length == 0) return;

    auto *rVB = _vbArr[_bufferPos];
    rVB->reset();
    rVB->writeBytes(reinterpret_cast<const char *>(_vb.getBuffer()), _vb.length());
    auto *uiMeshBuffer = _uiMeshBufferArr[_bufferPos];
    uiMeshBuffer->setVData(reinterpret_cast<float *>(rVB->getBuffer()));
    uiMeshBuffer->setByteOffset(static_cast<uint32_t>(_vb.length()));
}

void MeshBuffer::uploadIB() {
    auto length = _ib.length();
    if (length == 0) return;

    auto *rIB = _ibArr[_bufferPos];
    rIB->reset();
    if (rIB->getCapacity() < length) {
        rIB->resize(length, false);
    }
    rIB->writeBytes(reinterpret_cast<const char *>(_ib.getBuffer()), _ib.length());
    auto *uiMeshBuffer = _uiMeshBufferArr[_bufferPos];
    uiMeshBuffer->setIData(reinterpret_cast<uint16_t *>(rIB->getBuffer()));
}

void MeshBuffer::next() {
    _bufferPos++;
    if (_ibArr.size() <= _bufferPos) {
        auto *rIB = new IOTypedArray(se::Object::TypedArrayType::UINT16, _ib.getCapacity());
        _ibArr.push_back(rIB);
    }

    if (_vbArr.size() <= _bufferPos) {
        auto *rVB = new IOTypedArray(se::Object::TypedArrayType::FLOAT32, _vb.getCapacity());
        _vbArr.push_back(rVB);
    }

    if (_uiMeshBufferArr.size() <= _bufferPos) {
        addUIMeshBuffer();
    }
}

void MeshBuffer::reset() {
    _bufferPos = 0;
    _vb.reset();
    _ib.reset();
}

void MeshBuffer::addUIMeshBuffer() {
    UIMeshBuffer *uiMeshBuffer = new UIMeshBuffer();
    ccstd::vector<gfx::Attribute> attrs;
    if (_vertexFormat == VF_XYZUVC) {
        attrs = ATTRIBUTES_V3F_T2F_C4B;
    } else {
        attrs = ATTRIBUTES_V3F_T2F_C4B_C4B;
    }
    uiMeshBuffer->initialize(std::move(attrs), true);
    _uiMeshBufferArr.push_back(uiMeshBuffer);
}

void MeshBuffer::cleanUIMeshBuffer() {
    for (auto *buf : _uiMeshBufferArr) {
        delete buf;
    }
    _uiMeshBufferArr.clear();
}

cc::UIMeshBuffer *MeshBuffer::getUIMeshBuffer() const {
    return _uiMeshBufferArr[_bufferPos];
}

const ccstd::vector<cc::UIMeshBuffer *> &MeshBuffer::uiMeshBuffers() const {
    return _uiMeshBufferArr;
}

MIDDLEWARE_END

File: cocos/core/renderer/ui/mesh-buffer.ts
/*
 Copyright (c) 2019 Xiamen Yaji Software Co., Ltd.

 http://www.cocos.com

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated engine source code (the "Software"), a limited,
 worldwide, royalty-free, non-assignable, revocable and non-exclusive license
 to use Cocos Creator solely to develop games on your target platforms. You shall
 not use Cocos Creator software for developing other software or tools that's
 used for developing games. You are not granted to publish, distribute,
 sublicense, and/or sell copies of Cocos Creator.

 The software or tools in this License Agreement are licensed, not sold.
 Xiamen Yaji Software Co., Ltd. reserves all rights not expressly granted to you.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
*/

/**
 * @category ui
 */

import { GFXBuffer } from '../../gfx/buffer';
import { GFXBufferUsageBit, GFXMemoryUsageBit } from '../../gfx/define';
import { IGFXAttribute } from '../../gfx/input-assembler';
import { UI } from './ui';

export class MeshBuffer {
    public static OPACITY_OFFSET = 8;
    public batcher: UI;

    public vData: Float32Array | null = null;
    public iData: Uint16Array | null = null;

    public attributes: IGFXAttribute[] = null!;
    public vertexBuffers: GFXBuffer[] = [];
    public indexBuffer?: GFXBuffer;

    public byteStart = 0;
    public byteOffset = 0;
    public indicesStart = 0;
    public indicesOffset = 0;
    public vertexStart = 0;
    public vertexOffset = 0;
    public lastByteOffset = 1;

    public dirty = false;

    // NOTE:
    // actually 256 * 4 * (vertexFormat._bytes / 4)
    // include pos, uv, color in ui attributes
    private _vertexFormatBytes = 9 * Float32Array.BYTES_PER_ELEMENT;
    private _initVDataCount = 256 * this._vertexFormatBytes;
    private _initIDataCount = 256 * 6;
    private _outOfCallback: ((...args: number[]) => void) | null = null;

    constructor (batcher: UI) {
        this.batcher = batcher;
    }

    public initialize (attrs: IGFXAttribute[], outOfCallback: ((...args: number[]) => void) | null) {
        this._outOfCallback = outOfCallback;
        const vbStride = Float32Array.BYTES_PER_ELEMENT * 9;

        if (!this.vertexBuffers.length) this.vertexBuffers.push(this.batcher.device.createBuffer({
            usage: GFXBufferUsageBit.VERTEX | GFXBufferUsageBit.TRANSFER_DST,
            memUsage: GFXMemoryUsageBit.HOST | GFXMemoryUsageBit.DEVICE,
            size: vbStride,
            stride: vbStride,
        }));

        const ibStride = Uint16Array.BYTES_PER_ELEMENT;

        if (!this.indexBuffer) this.indexBuffer = this.batcher.device.createBuffer({
            usage: GFXBufferUsageBit.INDEX | GFXBufferUsageBit.TRANSFER_DST,
            memUsage: GFXMemoryUsageBit.HOST | GFXMemoryUsageBit.DEVICE,
            size: ibStride,
            stride: ibStride,
        });

        this.attributes = attrs;

        this._reallocBuffer();
    }

    public request (vertexCount = 4, indicesCount = 6) {
        this.lastByteOffset = this.byteOffset;
        const byteOffset = this.byteOffset + vertexCount * this._vertexFormatBytes;
        const indicesOffset = this.indicesOffset + indicesCount;

        if (vertexCount + this.vertexOffset > 65535) {
            // merge last state
            this.batcher.autoMergeBatches();
            if (this._outOfCallback) {
                this._outOfCallback.call(this.batcher, vertexCount, indicesCount);
            }
            return false;
        }

        let byteLength = this.vData!.byteLength;
        let indicesLength = this.iData!.length;
        if (byteOffset > byteLength || indicesOffset > indicesLength) {
            while (byteLength < byteOffset || indicesLength < indicesOffset) {
                this._initVDataCount *= 2;
                this._initIDataCount *= 2;

                byteLength = this._initVDataCount * 4;
                indicesLength = this._initIDataCount;
            }

            this._reallocBuffer();
        }

        this.vertexOffset += vertexCount;
        this.indicesOffset += indicesCount;
        this.byteOffset = byteOffset;

        this.dirty = true;
        return true;
    }

    public reset () {
        this.byteStart = 0;
        this.byteOffset = 0;
        this.indicesStart = 0;
        this.indicesOffset = 0;
        this.vertexStart = 0;
        this.vertexOffset = 0;
        this.lastByteOffset = 0;

        this.dirty = false;
    }

    public destroy () {
        this.attributes = null!;

        this.vertexBuffers[0].destroy();
        this.vertexBuffers.length = 0;

        if (this.indexBuffer) {
            this.indexBuffer.destroy();
            this.indexBuffer = undefined;
        }
    }

    public uploadData () {
        if (this.byteOffset === 0 || !this.dirty) {
            return;
        }

        const verticesData = new Float32Array(this.vData!.buffer, 0, this.byteOffset >> 2);
        const indicesData = new Uint16Array(this.iData!.buffer, 0, this.indicesOffset);

        if (this.byteOffset > this.vertexBuffers[0].size) {
            this.vertexBuffers[0].resize(this.byteOffset);
        }
        this.vertexBuffers[0].update(verticesData);

        if (this.indicesOffset * 2 > this.indexBuffer!.size) {
            this.indexBuffer!.resize(this.indicesOffset * 2);
        }
        this.indexBuffer!.update(indicesData);
    }

    private _reallocBuffer () {
        this._reallocVData(true);
        this._reallocIData(true);
    }

    private _reallocVData (copyOldData: boolean) {
        let oldVData;
        if (this.vData) {
            oldVData = new Uint8Array(this.vData.buffer);
        }

        this.vData = new Float32Array(this._initVDataCount);

        if (oldVData && copyOldData) {
            const newData = new Uint8Array(this.vData.buffer);
            for (let i = 0, l = oldVData.length; i < l; i++) {
                newData[i] = oldVData[i];
            }
        }
    }

    private _reallocIData (copyOldData: boolean) {
        const oldIData = this.iData;

        this.iData = new Uint16Array(this._initIDataCount);

        if (oldIData && copyOldData) {
            const iData = this.iData;
            for (let i = 0, l = oldIData.length; i < l; i++) {
                iData[i] = oldIData[i];
            }
        }
    }
}
